File: Python/lib/local_list.py
#!/usr/bin/env python
# -*- encoding: utf-8 -*-

import s3sync_lib
import os
import codecs
import re

config = s3sync_lib.get_config()
files = s3sync_lib.get_files()
output_file = os.path.dirname(__file__) + '/../working/LocalList.txt'

# Clear file
open(output_file, 'w').close()

for file in files:
    dir_path = os.path.abspath(file[u'path'])
    exclude = []
    rrs = config[u'rrs']
    encrypt = config[u'encrypt']
    alias = u''

    if u'rrs' in file:
        rrs = file[u'rrs']

    if u'encrypt' in file:
        encrypt = file[u'encrypt']

    if u'alias' in file:
        alias = file[u'alias']

    if u'exclude' in file:
        exclude = file[u'exclude']

    with codecs.open(output_file,encoding='UTF-8',mode='a') as lfile:
        dir_name = os.path.basename(dir_path)
        print('Listing files in %s...' % dir_name)

        for root,dir,files in os.walk(dir_path):
            for file in files:
                skip = False
                file_path = os.path.join(root,file)
                file_size = os.path.getsize(file_path)
                file_mtime = os.path.getmtime(file_path)

                for restr in exclude:
                    if re.match(restr,file):
                        skip = True
                        break

                if skip == True:
                    continue

                if file_size == 0:
                    continue

                lfile.write('<>'.join([
                    dir_name,
                    file_path,
                    str(file_size),
                    str(file_mtime),
                    str(rrs),
                    str(encrypt),
                    str(alias)
                ]))

                lfile.write(os.linesep)

File: PHP/lib/LocalList.php
<?php

require_once dirname(__FILE__) . '/./common.php';

$config = getConfig();
$files = getFiles();
$handle = fopen(dirname(__FILE__) . '/../working/LocalList.txt','w');

function recurse_directory($it,$file,$rrs,$encrypt) {
    global $handle;
    $dirsub = $file['dir'];
    $excludes = array();
    $alias = "";

    if(isset($file['exclude'])) {
        $excludes = $file['exclude'];
    }

    if(isset($file['alias'])) {
        $alias = $file['alias'];
    }

    try {
        while($it->valid()) {
            if($it->hasChildren()) {
                recurse_directory($it->getChildren(),$file,$rrs,$encrypt);
            } else if(!$it->isDot()) {
                $key = $it->key();
                $valid = true;

                foreach($excludes as $e) {
                    $check = "%$e%";

                    if(substr($e,0,1) === '/') {
                        $check = '%' . $dirsub . $e . '%';
                    }

                    if(preg_match($check,$key)) {
                        $valid = false;
                        break;
                    }
                }

                $size = filesize($key);
                if($valid && $size > 0) {
                    fwrite($handle,$dirsub . "<>$key<>" . $size . '<>' . filemtime($key) . "<>$rrs<>$encrypt<>$alias" .PHP_EOL);
                }
            }

            $it->next();
        }
    } catch(Exception $e) {
        throw $e;
    }
}

try {
    foreach($files as $file) {
        $rrs = $config['rrs'];
        $encrypt = $config['encrypt'];

        $it = new RecursiveDirectoryIterator(
            $file['path'],
            FilesystemIterator::UNIX_PATHS
        );

        if(isset($file['rrs'])) {
            $rrs = $file['rrs'];
        }

        if(isset($file['encrypt'])) {
            $encrypt = $file['encrypt'];
        }

        recurse_directory($it,$file,$rrs,$encrypt);
    }
} catch(Exception $e) {
    echo "Error: " . $e->getMessage();
    exit(1);
}

fclose($handle);
